### PR TITLE
MDL-72321 core_question: Implement question filtering

### PR DESCRIPTION
**Pull request by tuanngocnguyen**
This implementation will introduce question filtering
using the de-cupled participant filter. It makes changes
to the api to implement ajax based filtering as well as
the quiz to accept those filtering criteria.

Co-Authored-By: Safat Shahin <safatshahin@catalyst-au.net>
Co-Authored-By: Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
Co-Authored-By: Nathan Nguyen <nathannguyen@catalyst-au.net>

MDL-72321 core_question: Implement question filtering

This implementation will introduce question filtering
using the de-cupled participant filter. It makes changes
to the api to implement ajax based filtering as well as
the quiz to accept those filtering criteria.

Co-Authored-By: Safat Shahin <safatshahin@catalyst-au.net>
Co-Authored-By: Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
Co-Authored-By: Nathan Nguyen <nathannguyen@catalyst-au.net>

MDL-72752 mod_quiz: fix pagination and sorting on add question from question bank

MDL-72752 mod_quiz: add custom view for random question

WIP filterable object

 filter description

*** PLEASE DO NOT OPEN PULL REQUESTS VIA GITHUB ***

The moodle.git repository at Github is just a mirror of the official repository. We do not accept pull requests at Github.

See CONTRIBUTING.txt guidelines for how to contribute patches for Moodle. Thank you.

--

### CODE REVIEW

File: mod/quiz/amd/src/update_random_question_filter_condition.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+               window.location.href = returnUrl;
+            })
+           .catch(Notification.exception);
+    });
+
+    cancelButton.addEventListener("click", () => {
+        window.location.href = returnUrl;
+    });
+
+};
```
Comment (safatshahin) [inline mod/quiz/amd/src/update_random_question_filter_condition.js]: space issue here

File: mod/quiz/classes/external/add_random_question.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
+<?php
+
```
Comment (safatshahin) [inline mod/quiz/classes/external/add_random_question.php:16]: doc needed here

File: mod/quiz/classes/external/add_random_question.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            $categoryid = $qcobject->add_category($parentcategory, $newcategory, '', true);
+        }
+
+        // Add random question to the quiz.
+        list($quiz, ) = get_module_from_cmid($cmid);
+        quiz_add_random_questions($quiz, $addonpage, $categoryid, $randomcount, $filtercondition);
+        quiz_delete_previews($quiz);
+        quiz_update_sumgrades($quiz);
+
+        return ['message' => get_string('addarandomquestion_success', 'mod_quiz')];
```
Comment (safatshahin) [inline mod/quiz/classes/external/add_random_question.php:129]: is this used in ui? cant see that as a user, please lemme know if I am missing something

File: mod/quiz/classes/external/add_random_question.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+     * Returns description of method result value.
+     *
+     * @return external_value
+     */
+    public static function execute_returns() {
+        return new external_single_structure([
+            'message' => new external_value(PARAM_TEXT, 'Message', VALUE_OPTIONAL)
+        ]);
+    }
+}
```
Comment (safatshahin) [inline mod/quiz/classes/external/add_random_question.php]: space issue here

File: mod/quiz/classes/external/update_filter_condition.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
+<?php
+
```
Comment (safatshahin) [inline mod/quiz/classes/external/update_filter_condition.php:16]: doc here

File: mod/quiz/classes/external/update_filter_condition.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        // Validate context.
+        $thiscontext = \context_module::instance($cmid);
+        self::validate_context($thiscontext);
+        require_capability('mod/quiz:manage', $thiscontext);
+
+        // Update filter condition.
+        $DB->set_field('question_set_references', 'filtercondition', $filtercondition, ['id' => $id]);
+
+        return ['message' => get_string('updatefilterconditon_success', 'mod_quiz')];
```
Comment (safatshahin) [inline mod/quiz/classes/external/update_filter_condition.php:84]: same as above, cant see in ui

File: mod/quiz/classes/external/update_filter_condition.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+     * Returns description of method result value.
+     *
+     * @return external_value
+     */
+    public static function execute_returns() {
+        return new external_single_structure([
+            'message' => new external_value(PARAM_TEXT, 'Message', VALUE_OPTIONAL)
+        ]);
+    }
+}
```
Comment (safatshahin) [inline mod/quiz/classes/external/update_filter_condition.php]: space issue here

File: mod/quiz/classes/question/bank/custom_view.php [hunk 8]
```diff
@@ -198,68 +207,4 @@ protected function create_new_question_form($category, $canadd): void {
     protected function display_question_bank_header(): void {
     }
 
-    protected function build_query(): void {
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/custom_view.php]: why is this removed? this will introduce regression where quiz should only use ready questions not draft or hidden

File: mod/quiz/classes/question/bank/custom_view.php [hunk 9]
```diff
@@ -198,68 +207,4 @@ protected function create_new_question_form($category, $canadd): void {
...
-            }
-        }
-        // Build the SQL.
-        $sql = ' FROM {question} q ' . implode(' ', $joins);
-        $sql .= ' WHERE ' . implode(' AND ', $tests);
-        $this->countsql = 'SELECT count(1)' . $sql;
-        $this->loadsql = 'SELECT ' . implode(', ', $fields) . $sql . ' ORDER BY ' . implode(', ', $sorts);
-    }
-
-    public function add_standard_searchcondition(): void {
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/custom_view.php]: why is this removed? this will introduce regression where quiz should only use ready questions not draft or hidden

File: mod/quiz/classes/question/bank/random_question_view.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Defines the custom question bank view used on the Edit quiz page.
+ *
+ * @package   mod_quiz
+ * @category  question
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/random_question_view.php]: no need this doc

File: mod/quiz/classes/question/bank/random_question_view.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+ * @copyright 1999 onwards Martin Dougiamas and others {@link http://moodle.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace mod_quiz\question\bank;
+
+/**
+ * Subclass to customise the view of the question bank for the quiz editing screen.
+ *
+ * @author     2022 Nathan Nguyen <nathannguyen@catalyst-au.net>
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/random_question_view.php:23]: add the copyright here

File: mod/quiz/classes/question/bank/random_question_view.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * Prints the table of questions in a category with interactions
+     */
+    protected function display_question_list(): void {
+        echo \html_writer::start_tag('div',
+            ['class' => 'categoryquestionscontainer', 'id' => 'questionscontainer']);
+        echo \html_writer::end_tag('div');
+    }
+}
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/random_question_view.php]: space issue here

File: mod/quiz/classes/question/bank/random_question_view.php [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            'question_name_text_column',
+            'preview_action_column'
+        ];
+    }
+
+    /**
+     * Prints the table of questions in a category with interactions
+     */
+    protected function display_question_list(): void {
+        echo \html_writer::start_tag('div',
```
Comment (safatshahin) [inline mod/quiz/classes/question/bank/random_question_view.php]: can we use a simple mustache instead?

File: question/type/random/lang/en/qtype_random.php [hunk 14]
```diff
@@ -29,7 +29,7 @@
 $string['pluginname_help'] = 'A random question is not a question type as such, but is a way of inserting a randomly-chosen question from a specified category into an activity.';
 $string['pluginnameediting'] = 'Editing a random question';
 $string['privacy:metadata'] = 'The Random question type plugin does not store any personal data.';
-$string['randomqname'] = 'Random ({$a})';
+$string['randomqname'] = 'Random question based on filter condition';
```
Comment (safatshahin) [inline question/type/random/lang/en/qtype_random.php]: why this is changed? random qtype is not used anymore

File: question/type/random/questiontype.php [hunk 15]
```diff
@@ -28,6 +28,7 @@
 
 require_once($CFG->dirroot . '/question/type/questiontypebase.php');
 
+use core_question\local\bank\condition;
```
Comment (safatshahin) [inline question/type/random/questiontype.php]: why this is changed? random qtype is not used anymore

File: question/type/random/questiontype.php [hunk 16]
```diff
@@ -130,59 +131,12 @@ public function get_question_options($question) {
     /**
      * Random questions always get a question name that is Random (cateogryname).
      * This function is a centralised place to calculate that, given the category.
-     * @param stdClass $category the category this question picks from. (->parent, ->name & ->contextid are used.)
-     * @param bool $includesubcategories whether this question also picks from subcategories.
-     * @param string[] $tagnames Name of tags this question picks from.
+     * @param stdClass $slotdata the category this question picks from. (->parent, ->name & ->contextid are used.)
      * @return string the name this question should have.
      */
-    public function question_name($category, $includesubcategories, $tagnames = []) {
```
Comment (safatshahin) [inline question/type/random/questiontype.php]: why this is changed? random qtype is not used anymore, if this method is used, this is wrong, this plugin is getting removed in 4.1

File: question/type/random/questiontype.php [hunk 17]
```diff
@@ -130,59 +131,12 @@ public function get_question_options($question) {
...
-                return explode(',', $tagname)[1];
-            }, $tagnames));
-        } else {
-            $a = $categoryname ? : null;
-        }
-
-        $name = get_string($stringid, 'qtype_random', $a);
-
-        return shorten_text($name, 255);
+    public function question_name($slotdata) {
```
Comment (safatshahin) [inline question/type/random/questiontype.php]: why this is changed? random qtype is not used anymore

**Review by safatshahin — CHANGES_REQUESTED** [review]
Hi @tuanngocnguyen 
Thank you for working on the patch. Patch looks really good, I made some comments about quiz custom view where it might introduce a regression. There have been changes in random qtype, which is concerning as this qtype is not used anymore from 4.0, please have a look.
Please also address the failed unit tests, we can address the behats using a different MR.
As soon as they are addressed, this branch is good to go with the base branch.
Cheers!

File: mod/quiz/classes/external/add_random_question.php (hunk 3 above)
Comment (tuanngocnguyen) [reply mod/quiz/classes/external/add_random_question.php:129]: The ws function is used in this js: https://github.com/catalyst/moodle-MDL-72752/pull/63/files#diff-ad78a10007735f9361ff037bfec8321d5502d65036d064c6a3628692b3cc4729R334

File: mod/quiz/classes/external/update_filter_condition.php (hunk 6 above)
Comment (tuanngocnguyen) [reply mod/quiz/classes/external/update_filter_condition.php:84]: It is used in this js: https://github.com/catalyst/moodle-MDL-72752/pull/63/files#diff-404b42454a1a8069c4d1790521c736773515d6b7cdf7fe160f8ae3d5a0dca483R46

File: mod/quiz/classes/question/bank/custom_view.php (hunk 8 above)
Comment (tuanngocnguyen) [reply mod/quiz/classes/question/bank/custom_view.php]: I will review and add it back

File: mod/quiz/classes/question/bank/random_question_view.php (hunk 13 above)
Comment (tuanngocnguyen) [reply mod/quiz/classes/question/bank/random_question_view.php]: I will add one

File: mod/quiz/edit.php [hunk 18]
```diff
@@ -131,15 +130,33 @@
...
+            $defaultcategoryobj->id,
+            $defaultcategory,
+            null,
+            $contexts->having_cap('moodle/question:add'));
+        $categoryname = required_param('categoryname', PARAM_TEXT);
+        $parentcategory = required_param('parentcategory', PARAM_TEXT);
+        $categoryid = $qcobject->add_category($parentcategory, $categoryname, '', true);
+    }
+
+    quiz_add_random_questions($quiz, $addonpage, $categoryid, $randomcount, $recurse);
```
Comment (tuanngocnguyen) [inline mod/quiz/edit.php]: I forget removing this one, as I have replaced it with ws function mod_quiz_add_random_question

File: mod/quiz/classes/external/add_random_question.php (hunk 3 above)
Comment (safatshahin) [reply mod/quiz/classes/external/add_random_question.php:129]: correct, but why the return message? the return message seems not in use

Comment (tuanngocnguyen) [reply mod/quiz/classes/external/add_random_question.php:129]: I have updated with messages are used in https://github.com/catalyst/moodle-MDL-72752/pull/63/commits/fc6dea0aef23893438f9ef3da241f745d296077b#diff-9cdb99b99ca28cfcf036270f6b963eb974353a52945cffb6ae4752454d6702b9R173

**Comment by tuanngocnguyen** [comment]
Hi @safatshahin ,

I have updated the MR with fixed issues from the previous review, unit test, history plugin and rebase. 
Please review and let me know if there is anything I need to fix
Thanks

**Comment by safatshahin** [comment]
Hi @tuanngocnguyen 
The patch looks good with the changes so far, we need to think about the restore bit but that can be done differently.
The big problem with this patch is issues with commits and MR changes.
There have been a lot of changes in the filtering base, for ex, regression fix, behat fix, etc, whatever is fixed, these commits seems to remove them as a part of the MR as your commits don't have them. I can see it's almost impossible to track what you changed and what the Mr is merging as a part of the commit as the commits are mixed and hashes don't match.
What you will have to do is, bring your commits specifically for these changes, rebase against the filter base branch and then do an MR.
Cheers

**Comment by safatshahin** [comment]
@tuanngocnguyen 
The patch looks good, we also need some unit test coverage for backup and restore of random question filter conditions as well as some scenario-based test coverage. Those can be covered in a different git issue.
Cheers
